src/runtime_src/core/tools/xbflash2/OO_Program_Qspips.cpp

```diff
@@ -0,0 +1,295 @@
+#include "tools/common/XBUtilitiesCore.h"
+
+#include "core/pcie/tools/xbflash.qspi/firmware_image.h"
+#include "core/pcie/tools/xbflash.qspi/pcidev.h"
+#include "core/pcie/tools/xbflash.qspi/xqspips.h"
+
+namespace XBU = XBUtilities;
+
+// 3rd Party Library - Include Files
+#include <boost/program_options.hpp>
+#include <boost/format.hpp>
+#include <boost/algorithm/string.hpp>
+namespace po = boost::program_options;
+
+// System - Include Files
+#include <iostream>
+#include <fstream>
+
+namespace {
+
+int 
+qspips_flash(po::variables_map& vm) {
+    std::string bdf;
+    std::string flash_type;
+    std::string bin_file;
+    std::vector<std::string> bin_files;
+    std::string soffset;
+    size_t offset = 0;
+    bool force = false;
+    std::string sBar, sBarOffset;
+    size_t baroff = INVALID_OFFSET;
+    int bar = 0;   
+
+    XBU::sudo_or_throw_err();
+     
+    //mandatory command line args
+    try {
+        bdf = vm["device"].as<std::string>();
+        bin_files = vm["image"].as<std::vector<std::string>>();
+    }
+    catch (...) {
+        throw std::errc::invalid_argument;
+    }
+
+    //optional command line args
+    try {
+        if (vm.count("flash-part"))
+            flash_type = vm["flash-part"].as<std::string>();
+        if (vm.count("offset")) {
+            soffset = vm["offset"].as<std::string>();
+            if (!soffset.empty()) {
+                std::stringstream sstream(soffset);
+                sstream >> offset;
+            }
+        }
+        if (vm.count("force"))
+            force = true;
+    }
+    catch (...) {
+    }
+
+    //optionals command line args
+    try {
+        if (vm.count("bar")) {
+            sBar = vm["bar"].as<std::string>();
+            if (!sBar.empty())
+                bar = std::stoi(sBar);
+        }
+        if (vm.count("bar-offset")) {
+            sBarOffset = vm["bar-offset"].as<std::string>();
+            if (!sBarOffset.empty()) {
+                std::stringstream sstream(sBarOffset);
+                sstream >> baroff;
+            }
+        }
+    }
+    catch (...) {
+    }
+
+    if (bdf.empty()) {
+        std::cout << "Error: Please provide mgmt BDF\n";
+        throw std::errc::invalid_argument;
+    }
+
+    if (!bin_files.size()) {
+        std::cout << "Error: Please provide proper BIN file.\n";
+        throw std::errc::invalid_argument;
+    }
+
+    firmwareImage bin(bin_files[0].c_str());
+    if (bin.fail()) {
+        std::cout << "Error: Please provide proper BIN file.\n";
+        throw std::errc::invalid_argument;
+    }
+    std::cout << "About to program flash on device "
+        << boost::format(" %s at offset 0x%x\n") % bdf % offset;
+
+    if (!force && !XBU::can_proceed())
+        throw std::errc::operation_canceled;
+
+    pcidev::pci_device dev(bdf, bar, baroff, flash_type);
+    XQSPIPS_Flasher qspips(&dev);
+
+    return qspips.xclUpgradeFirmware(bin, offset);
+}
+
+int 
+qspips_erase(po::variables_map& vm) {
```

Comment:
This function looks almost the same as qspips_flash().  Please consider refactoring the code to a common helper function.